api/authentication/access_token.py

```diff
@@ -0,0 +1,201 @@
+from __future__ import annotations
+
+import logging
+import time
+from datetime import timedelta
+from typing import Optional, Type, cast
+
+from jwcrypto import jwe, jwk
+
+from api.authentication.base import AuthProviderSettings
+from api.problem_details import (
+    PATRON_AUTH_ACCESS_TOKEN_EXPIRED,
+    PATRON_AUTH_ACCESS_TOKEN_INVALID,
+)
+from core.integration.goals import Goals
+from core.integration.settings import ConfigurationFormItem, FormField
+from core.model import get_one_or_create
+from core.model.configuration import ExternalIntegration
+from core.model.integration import IntegrationConfiguration
+from core.model.patron import Patron
+from core.util.datetime_helpers import utc_now
+from core.util.problem_detail import ProblemDetail, ProblemError
+from core.util.string_helpers import random_string
+
+
+class PatronAccessTokenProvider:
+    """Provides access tokens for patron auth"""
+
+    @classmethod
+    def generate_token(
+        cls, _db, patron: Patron, password: str, expires_in: int = 3600
+    ) -> str:
+        raise NotImplementedError()
+
+    @classmethod
+    def decode_token(cls, _db, token: str) -> dict | ProblemDetail:
+        raise NotImplementedError()
+
+    @classmethod
+    def is_access_token(cls, token: str | None) -> bool:
+        raise NotImplementedError()
+
+    @classmethod
+    def get_integration(cls, _db):
+        raise NotImplementedError()
+
+
+class PatronAccessTokenAuthenticationSettings(AuthProviderSettings):
+    auth_key: Optional[str] = FormField(
+        None,
+        description="The JWE key used for the access token",
+        form=ConfigurationFormItem("Auth Key"),
+    )
+
+
+class PatronJWEAccessTokenProvider(PatronAccessTokenProvider):
+    """Provide JWE based access tokens for patron auth"""
+
+    NAME = "Patron Access Token Provider"
+
+    @classmethod
+    def generate_key(cls) -> jwk.JWK:
+        """Generate a new key compatible with the token encyption type"""
+        kid = random_string(16)
+        return jwk.JWK.generate(kty="oct", size=256, kid=kid)
+
+    @classmethod
+    def get_integration(cls, _db) -> IntegrationConfiguration:
+        integration, _ = get_one_or_create(
+            _db,
+            IntegrationConfiguration,
+            protocol=ExternalIntegration.PATRON_AUTH_JWE,
+            goal=Goals.PATRON_AUTH_GOAL,
+            name=cls.NAME,
+        )
+        return integration
+
+    @classmethod
+    def rotate_key(cls, _db) -> jwk.JWK:
+        """Rotate the current JWK key in the DB"""
+        key = cls.generate_key()
+        integration = cls.get_integration(_db)
+
+        # We must make a copy() because sqlalchemy does not think
+        # a change in the same dict is a change at all.
+        # So the dict must be a new dict to register the change.
+        # https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSONB
+        settings_dict: dict = cast(dict, integration.settings.copy())
```

Comment:
@jonathangreen The fix was to have a completely new dict object as opposed to changing the current object

Reply:
The solution came while I was asking phind.com for random solutions :sweat_smile: . Time to be replaced :robot: